src/lib.rs

```diff
@@ -716,6 +692,9 @@ pub trait Log: Sync + Send {
     /// Implementations of `log` should perform all necessary filtering
     /// internally.
     fn log(&self, record: &Record);
+
+    /// Flushes any buffered records.
+    fn flush(&self) {}
```

Comment:
I think the issue mentioned a default impl to retain back compat, which we don't really need to worry about since we're removing shutdown at the same time.

I don't think we'd want to add the refcounting into the flush impl for env_logger - it'd defeat the purpose of getting rid of shutdown in the first place! The use case for flush is when you're for example, in the shutdown process of your application. If you use something like env_logger then just killing the process is fine - any log calls that finished before the exit will end up on the console. But, if you're using a logger that performs the IO asynchronously, you can end up in situations where log lines that should have been output aren't. Imagine something like

```rust
info!("shutting down");
log::flush();
process::exit(0);
```

All we need to make sure is that the "shutting down" message makes it all the way through the logging infrastructure. If there are other threads doing things in the background, the state of their messages is not super important since they're racing with this thread.